### PR TITLE
Some thoughts on workflow topics and queues

### PR DESCRIPTION
**Pull request by tomcrane**
Not quite an RFC but wanted to get some notes down somewhere other than Slack.

### CODE REVIEW

**Comment by tomcrane** [comment]
I wonder how it got docs/getting-started.md... I can kinda see how that might have happened but..

File: docs/workflow-messaging.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+- born-digital-bag-notifications-staging
+
+There will be two new SNS topics for Goobi:
+
+- digitised-bag-notifications-prod
+- digitised-bag-notifications-staging
+
+_(I assume, or similar names anyway)_
+
+There are/will be queues subscribed to these topics. Where do the queues live? We have currently one production queue, known as:
```
Comment (donaldgray) [inline docs/workflow-messaging.md:15]: > Where do the queues live?

I'd prefer they live in digirati acct. Keeps all of the subscriber logic in one place and the publishing logic in another.

File: docs/workflow-messaging.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+arn:aws:sqs:eu-west-1:975596993436:born-digital-notifications-prod (in storage acct)
+arn:aws:sqs:eu-west-1:653428163053:born-digital-notifications-prod (in digirati acct)
```
Comment (donaldgray) [inline docs/workflow-messaging.md:18]: These 2 need to be bullets for formatting

File: docs/workflow-messaging.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+There is also similar for staging. 
+
+The WorkflowProcessor listens to just this queue at the moment, because we've only tested the one environment and it's only for born-digital events.
+
+Will we have another pair of queues for prod and staging that recieve messages from the digitised-bag-notifications-xxx topic? So that a running WorkflowProcessor is picking up from two queues?
+Or should we have one queue that subscribes to both born-digital and digitised topics, and gets workflow messages from both? (should rename that queue if that's the case).
```
Comment (donaldgray) [inline docs/workflow-messaging.md:25]: I think 2 queues as it's a lot simpler. 1 single queue wouldn't really work for pausing from one source.

We could use messageFiltering to pause from one origin but you'd lose delivery of the message. If we use the "open and look" then the paused-origin messages would eventually be sent to DLQ as the maxReceives would be hit fairly quick (and non-paused origin messages would increasingly be slow to process as there'd be lots of ones we're not interested in).

File: docs/workflow-messaging.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+- born-digital-bag-notifications-dev-stage-storage
+- digitised-bag-notifications-dev-stage-storage
+
+... to give us independent queues that won't be getting picked up by any deployed WorkflowProcessor, for all combinations we'd want to debug / investigate.
+This requires 4 queues being kept up that are hardly ever used, though. Maybe **not** have these but just have the dev ones as in the previous list but have two extra _topics_:
+
+- born-digital-bag-notifications-dev
+- digitised-bag-notifications-dev
+
+Neither Goobi nor the storage service broadcast to these topics, we have to manually send to them, for whatever dev setup we are testing to pick up.
```
Comment (donaldgray) [inline docs/workflow-messaging.md:76]: I prefer this approach of having to manually publish to SNS topics for dev. SNS + SQS can all live in digrati account.

File: docs/workflow-messaging.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+Either way the dashboard will need to pick the right topic to notify when manually running an end-to-end workflow from the dashboard, pretending to be Goobi or the storage service (it puts "dashboard" in the Origin of the [message](https://github.com/wellcomecollection/iiif-builder/blob/develop/src/Wellcome.Dds/Wellcome.Dds.Common/WorkflowMessage.cs) to be clear where the message came from).
+
+We can have more than one WorkflowProcessor service picking up a job, but we don't want more than one environment synchronising with the same DLCS space.
+So DDS production and test each have their own queue(s) subscribed to both Goobi and born-digital topics; prod writes to space 5 and test writes to space 7 (this is NEW)
+DDS staging writes to DLCS space 6, as now.
+
+In normal operation the `Test` environment does not listen to queues, or it throws away messages, etc - we control this through a flag so we can turn test on went we want to observe it. Even when test is running, its workflowProcessor is not always acting on the messages it picks up; it needs a control switch (e.g., a sentinel S3 key) to act on the messages.
```
Comment (donaldgray) [inline docs/workflow-messaging.md:35]: > e.g., a sentinel S3 key

Something that hooks into IOptions model/appSettings would be more idiomatic. Maybe something like parameterStore that could periodically refresh `IOptionsMonitor`.

**Review by donaldgray — APPROVED** [review]
Required SNS/SQS makes sense, added any comments inline.

**Pull request merged**